service/service.go

```diff
@@ -91,23 +92,41 @@ type DeployStatus struct {
 
 // New creates a new service from a gzipped tarball.
 func New(tarball io.Reader, env map[string]string, options ...Option) (*Service, error) {
+	var err error
 	s := &Service{}
-
 	defer s.closeStatus()
 
 	if err := s.setOptions(options...); err != nil {
 		return nil, err
 	}
-	if err := s.saveContext(tarball); err != nil {
+
+	// untar tarball to retrieve mesg.yml
+	s.tempPath, err = ioutil.TempDir("", "mesg-")
+	if err != nil {
 		return nil, err
 	}
-	defer os.RemoveAll(s.tempPath)
 
+	if err := archive.Untar(tarball, s.tempPath, nil); err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(s.tempPath)
 	def, err := importer.From(s.tempPath)
 	if err != nil {
 		return nil, err
 	}
 
+	// create stream to calculate hash
+	decompressedStream, err := archive.Tar(s.tempPath, archive.Uncompressed)
+	if err != nil {
+		return nil, err
+	}
+	envbytes := []byte(xos.EnvMapToString(env))
+	hash, err := tarsum.New(tar.NewReader(decompressedStream)).Sum(envbytes)
```

Comment:
Can't we use the reader that we have in parameter already, I don't see the point of having to tar it again except the `Uncompressed` but why this is necessary ? Can we have some details on that.

Also I think there is function for the hash that accepts a reader so why do we need to create this `tarsum` package ?

Reply:
> Can't we use the reader that we have in parameter already,
No because the reader can be only read once. Here we save context to extract the mesg.yaml, then calculate the tarsum.

> Also I think there is function for the hash that accepts a reader so why do we need to create this `tarsum` package ?

Can you point one you are talking about?

Reply:
https://github.com/mesg-foundation/core/pull/731/commits/c65fadc878c01e1f4680a0e86bfedcf3e148ec2f#diff-eeade751807b5e894d1f9a5b21390878

What's wrong with this implementation ?
This one plus the `mapEnvToString` looks valid for me why not go with that ? Am I missing something.

Reply:
The implementation is wrong:

- the hash won't be deterministic if the compression is different
- tar has [headers](https://golang.org/pkg/archive/tar/#Header) that may affect the hash calculation - for example, the mod/access/change times for a file (although the context of the fill will be the same)

Reply:
That's the part I don't really understand.

> the hash won't be deterministic if the compression is different

Agree but we control the compression when we deploy it and/or github have the same algorithm . Also the archive itself is a version. As long as this archive is the same for every computer it's fine (if we regenerate one and it's a different it's fine)

> tar has [headers](https://golang.org/pkg/archive/tar/#Header) that may affect the hash calculation

Again this is in the archive as long as we have the archive it will be the same except if when we untar it then it changes but this shouldn't be the case.

Reply:
> Also the archive itself is a version.
Hum? What version? the only `version` we're using is hash.

> . As long as this archive is the same for every computer it's fine

What do you mean by "the same for every computer"

> (if we regenerate one and it's different it's fine)

So if we regenerate archive and the content will be the same but the hash different then we don't have deterministic hash and this was the purpose of this PR, wasn't it?

> Again this is in the archive as long as we have the archive it will be the same except if when we untar it then it changes but this shouldn't be the case.

But you want to support multiple releases. Archive on github will be different, then one put on CI Artifacts then one put on ipfs. And there is also git repo which we tar ourselves. Someone might archive the same files on different os, machine, and the headers might be different.

Reply:
can you show me how they are different, for me from my previous tests all the version generated were the same: git, tar from github, tar from local source so if the generation is the same then the goal is to have a similar hash.
I just really don't see the extra work on this one when this technic works fine https://github.com/mesg-foundation/core/commit/c65fadc878c01e1f4680a0e86bfedcf3e148ec2f#diff-eeade751807b5e894d1f9a5b21390878. After maybe it works fine only on osx and in that case i'm fine but I would like to see and understand really why there is all this `tarsum` package that based on my tests is not necessary

Reply:
> can you show me how they are different, for me from my previous tests all the version generated were the same: git, tar from github, tar from local source so if the generation is the same then the goal is to have a similar hash.

There were too few tests and it happens to be the same. Here you have a showcase what I'm talking about: https://play.golang.org/p/a_rKlE0jk5t.

The content of the files are the same but the hash is different. Again there are tar headers and someone might build the same source multiple times on different os/machine/timezones etc. This is why we need tarsum to have exact hash.

Reply:
of course these are different, in your example you are setting a different modification time for each files
```go
ModTime: time.Date(i, 0, 0, 0, 0, 0, 0, time.UTC),
```
where `i` is either `0` or `1`.

If you force it to the same (for example `0`) you have a consistent hash `8c60defcbc9ff6ce751b290dcf938c8c538e9028`

And in a real life application if the tar changes because of a modification on date of a file or the permission or whatever metadata/header of a file then it's a new version and a it's normal not to have the same hash.

I would love the feedback from @NicolasMahe and @ilgooz on that but your example exactly shows that with the same environment the hash is the same and yes if we change the environment the hash changes but that's exactly the idea.

So in that case all the tarsum package is irrelevant.

Reply:
Yes and this is a real situation. You could have multiple ci pipelines for the same build that for example regenerates some scripts/source code and change the mod/access time etc and in that case, the version stays the same. Thinking that there will be only one tar for the version is an assumption that may not be true in some cases.

Reply:
The old idea behind of announcing hash sums for tar files are actually for security. This is the pattern we see always. When program makers publish a tarball via a 3rd party, they also announce the hash sum for that tarball so people can recalculate it from the downloaded file and verify the truth of untouched tarball.

In the normal sense, users can simply verify a tarball that way. But this is only possible by creating tarball once and distributing copies from that one. This is a very simplistic way of validation or can be considered as a limitation on requiring to create the tarball only once, depending on how you look at it.

In our case, we're trying to support git, url tarballs and ipfs together. Which means we end up 3 different tarballs generated. This way we have the flexibility of supporting tarballs generated from different sources or even with different algorithms but it's not simplistic anymore on the validation side. And we have env variable values that needs to change the result of the calculations which brings even more complexity. 

@antho1404 as you mentioned, either we should threat this 3 tarballs as different services because their tarball calculations are different or we need to create our own calculation strategy like @krhubert did with this PR. Or another option is might be to give devs to possibility to provide multiple tarball sources at the marketplace publish time. So we can download and calculate hashes and put them into service manifest as aliases of each other. But still it's complicated to have different hashes for the same service.

Also when we bring env vars in to the table while calculating the hashes, it still complicates stuff because we're creating a custom hash calculation strategy anyway. This might be ok. But getting rid of that is also simple. And I'm a bit on that side as well. It's also odd to create a new **static** service hash depending on the configuration values like env vars. Because the env var values are generally application specific and this new hash(service) generated from the env var values shouldn't be published on the marketplace anyway. So there is no actual reason to apply env variables on the deploy time, I think.

We can apply them on start time. The only difference here is that we need to pass the env vars each time while starting a service and this is actually a good thing. No hidden configuration going on under the hood.

As a conclusion, I'd like to apply env variables to services at the start time and only respect to service hash that is in the service manifest saved to ipfs during the marketplace publish. Other deploy urls for the same service different than the one in the manifest should not be considered as legit nor shouldn't be profitable for the dev as in tokens.

Reply:
I think we are trying to solve 2 different problems with one solution that is complicated for the first problem and not "enough" for the second:

- The first problem to resolve is about verifying that the downloaded source code of a service is the one that the developer published.
This is a checksum system that could be **only** available on service downloaded from the marketplace. In this case, the source code is uploaded as an archive created once and its checksum will be always the same because there is no decompression (so modification of file headers) or whatever that come into the calculation of the checksum.
As the archive is currently uploaded to IPFS, its checksum is already created and verified by IPFS libs, so we might not even need to verify it again in the core.
This was the first problem and why the implemented in this PR is actually too complicated.

- The second problem to resolve is to guarantee that the exact same software built from the same source is actually the same.
This current implementation is not enough because it's not verifying ALL the dependencies and the version of the system (eg ubuntu vs debian) and the compiler version (node vX vs node vY, go vX vs go vY, etc..) and many other things...
In order to resolve this problem, the docker image should be built on the computer of the developer and fully uploaded to the marketplace (like Docker Hub). This way, not only the source code but also ALL the OS and the dependencies will be exactly the same.
That's why I said this implementation is not "enough" for the second problem.

----

I think the hash of a service should be the hash of the docker image and not the hash of the source files. So actually, **forgetting completely** about a deterministic hash that could solve both problems.

Reply:
I agree on that same service sources does not mean that a docker image built from that source from different environments will be the same. Because a docker image might be using _from x:latest_ tag or downloading other softwares during the build process while not explicitly fixing their versions.

With this awareness, if we really want the services to be truly same, we can start using the hash of Docker image instead of the service's source code as @NicolasMahe suggested and publish the image to ipfs.

But I'm not sure to go with this way because I'd like to give full control to devs about building their Docker images from the "true" source code. Because Docker images kind of black boxes and you'll not be easily know what is inside once it's built. Also, it is hard for devs to verify Docker image's hash if they want to for the same reason about unfixed versions.

So I think, we should give this responsibility to service authors about fixing their versions inside the Dockerfile and keep publishing service's tar file to ipfs and use its hash. And move env variables to start time and accept only one tar file url for the service from a single source.

Reply:
> I think the hash of a service should be the hash of the docker image and not the hash of the source files. So actually, forgetting completely about a deterministic hash that could solve both problems.

The docker hash image is not deterministic, we've checked this.

Reply:
> I think we are trying to solve 2 different problems with one solution that is complicated for the first problem and not "enough" for the second:

For me the problem is that we want for the same archive, the one from the marketplace to make sure that this will generate always the same hash.

I don't think we should even discuss your second problem yet. As for using the docker image as hash like we saw this is not enough. For me here the only stuff that is valid whatever the machine is doing the shaX sum of the sources (the tarball from the marketplace) I would not even think about the one from git or anything (even if like the initial commit shows, it's the same).

@NicolasMahe I suggest to close this PR if we are not clear on the problem then discuss the problem and then implement the solution. From my understanding it's having a same hash for a given tarball (that will not change).

Reply:
resolved here https://forum.mesg.com/t/services-hash/261/6?u=anthony